enhancements/installer/vsphere-ipi-zonal.md

```diff
@@ -279,24 +315,160 @@ type VSpherePlatformSpec struct {
+	// +listType=atomic
+	// +optional
+	APIServerInternalIPs []IP `json:"apiServerInternalIPs"`
+
+	// ingressIPs are the external IPs which route to the default ingress
+	// controller. The IPs are suitable targets of a wildcard DNS record used to
+	// resolve default route host names.
+	// In dual stack clusters this list contains two IP addresses, one from IPv4
+	// family and one from IPv6.
+	// In single stack clusters a single IP address is expected.
+	// When omitted, values from the status.ingressIPs will be used.
+	// Once set, the list cannot be completely removed (but its second entry can).
+	//
+	// +kubebuilder:validation:MaxItems=2
+	// +kubebuilder:validation:XValidation:rule="size(self) == 2 && isIP(self[0]) && isIP(self[1]) ? ip(self[0]).family() != ip(self[1]).family() : true",message="ingressIPs must contain at most one IPv4 address and at most one IPv6 address"
+	// +listType=atomic
+	// +optional
+	IngressIPs []IP `json:"ingressIPs"`
+
+	// machineNetworks are IP networks used to connect all the OpenShift cluster
+	// nodes. Each network is provided in the CIDR format and should be IPv4 or IPv6,
+	// for example "10.0.0.0/8" or "fd00::/8".
+	// +listType=atomic
+	// +kubebuilder:validation:MaxItems=32
+	// +kubebuilder:validation:XValidation:rule="self.all(x, self.exists_one(y, x == y))"
+	// +optional
+	MachineNetworks []CIDR `json:"machineNetworks"`
 }
 ```
 
-### CCCMO
+##### Updates required for vm-host group zonal
 
-The CCCMO cloud-config transformation already exists but will 
-need to be modified to support the VSpherePlatformSpec.
+```golang
 
-### installer
+// The VSphereFailureDomainType is a string representation of a failure domain
+// type. There are three supportable types: HostGroup, ComputeCluster and Datacenter
+type VSphereFailureDomainType string
+
+const (
+	// HostGroupFailureDomain is a failure domain for a vCenter vm-host group.
+	HostGroupFailureDomain VSphereFailureDomainType = "HostGroup"
+	// ComputeClusterFailureDomain is a failure domain for a vCenter compute cluster.
+	ComputeClusterFailureDomain VSphereFailureDomainType = "ComputeCluster"
+	// DatacenterFailureDomain is a failure domain for a vCenter datacenter.
+	DatacenterFailureDomain VSphereFailureDomainType = "Datacenter"
+)
+
+// VSpherePlatformFailureDomainSpec holds the region and zone failure domain and the vCenter topology of that failure domain.
+// +kubebuilder:validation:XValidation:rule="has(self.zoneType) && self.zoneType == 'HostGroup' ? self.topology.affinity.vmGroup != '' && self.topology.affinity.hostGroup != '' && self.topology.affinity.vmHostRule != '' : true",message="when zoneType is HostGroup, failuredomain topology affinity vmGroup, hostGroup and vmHostRule fields must be defined"
```

Comment:
Topology does not appear in the spec here, I think it may have been meant to be there though?

This validation apppears that you want an optional struct, with required fields in for the zoneType, is this just a discriminated union? It would then have the standard that `hostGroup` is allowed if and only if `type == HostGroup`

```
region:
  type: ABC
zone:
  type: HostGroup
  hostGroup:
    vmGroup: ...
    hostGroup: ...
    vmHostRule: ...
```

Reply:
I broke out what is current (as of 9/9/24) and what is specifically changing for vm-host zonal. Would you rather have that as a single section?

Reply:
personally - it makes sense to put new changes into existing structs, rather than breaking them out. This way, we can still see what changed via diff, but we can also see the full datatype if we view complete file.

Reply:
@JoelSpeed does this look like the direction you are thinking? https://github.com/openshift/enhancements/pull/1677/commits/677ce07864da82e457922281634b05e5c7c20523

Reply:
or something like this: https://github.com/openshift/enhancements/pull/1677/commits/fdb3dde5dc1d3ea6862c13e42e6e1a1339fbf67a


Since there is a name collision with region and zone, wondering if it should be moved to the topology struct

Reply:
The latter of the two commits is where I was thinking to go

Reply:
_whew_ thats what I ended up with 

https://github.com/openshift/api/commit/906bb28eac2277d71a11c35dded4578db38f6598

The enhancement has been updated for that

https://github.com/openshift/enhancements/pull/1677/commits/ea2162a215491d563646fb7b59a638ebf66c5dc4